File: Dynamic_Programming/problem22/problem22.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        dp: List[List[str]] = [[] for _ in range(n + 1)]
        dp[0] = [""]

        for k in range(1, n + 1):
            cur: List[str] = []
            for i in range(k):
                for left in dp[i]:
                    for right in dp[k - 1 - i]:
                        cur.append("(" + left + ")" + right)
            dp[k] = cur

        return dp[n]

File: Dynamic_Programming/problem22/problem22.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<vector<string>> dp(n + 1);
        dp[0] = {""};

        for (int k = 1; k <= n; ++k) {
            vector<string> cur;
            for (int i = 0; i < k; ++i) {
                for (const string& left : dp[i]) {
                    for (const string& right : dp[k - 1 - i]) {
                        cur.push_back("(" + left + ")" + right);
                    }
                }
            }
            dp[k] = move(cur);
        }

        return dp[n];
    }
};
